src/query/query_executor.c

```diff
@@ -11585,10 +11586,11 @@ qexec_execute_obj_fetch (THREAD_ENTRY * thread_p, XASL_NODE * xasl, XASL_STATE *
       scan_cache_end_needed = true;
 
       /* fetch the object and the class oid */
-      if (heap_get_with_class_oid (thread_p, &cls_oid, dbvaloid, &oRec, &scan_cache, scan_operation_type, PEEK,
-				   LOG_ERROR_IF_DELETED) != S_SUCCESS)
+      scan = heap_get_for_operation (thread_p, dbvaloid, &cls_oid, &oRec, &scan_cache, scan_operation_type, PEEK,
+				     LOG_ERROR_IF_DELETED);
+      if (scan != S_SUCCESS)
 	{
-	  if (er_errid () == ER_HEAP_UNKNOWN_OBJECT)
+	  if (scan == S_DOESNT_EXIST || scan == S_SNAPSHOT_NOT_SATISFIED || er_errid () == ER_HEAP_UNKNOWN_OBJECT)
```

Comment:
I think this is part of your idea to avoid relying on ER_HEAP_UNKNOWN_OBJECT error when deleted objects are expected. I think leaving a TODO comment explaining intended removal of er_errid () == ER_HEAP_UNKNOWN_OBJECT.